mailbox/noise.go

```diff
@@ -321,27 +345,391 @@ func newHandshakeState(initiator bool, prologue []byte,
+
+			if h.payloadToSend != nil {
+				var length [lengthHeaderSize]byte
+				payLoadLen := len(h.payloadToSend)
+				binary.BigEndian.PutUint16(
+					length[:], uint16(payLoadLen),
+				)
+
+				_, err := payloadWriter.Write(length[:])
+				if err != nil {
+					return err
+				}
+
+				_, err = payloadWriter.Write(h.payloadToSend)
+				if err != nil {
+					return err
+				}
+			}
+
+			copy(payload[:], payloadWriter.Bytes())
+
+		default:
+			return fmt.Errorf("unknown act number: %d", mp.ActNum)
+		}
+
+		_, err := buff.Write(h.EncryptAndHash(payload))
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("unknown handshake version: %v", h.version)
 	}
+
+	_, err := w.Write(buff.Bytes())
+	return err
 }
 
-// AuthDataPayload is a functional option that allows the non-initiator to
-// specify a piece of data that should be sent to the initiator during the
-// final phase of the handshake. This information is typically something like a
-// macaroon.
-func AuthDataPayload(authData []byte) func(*Machine) {
-	return func(m *Machine) {
-		m.authData = authData
+// readMsgPattern uses the given MessagePattern to process the data received
+// from the passed reader. First the handshake version is checked, then
+// the MessagePattern Tokens are processed and then any extra cipher text is
+// decrypted as defined by the handshake version.
+func (h *handshakeState) readMsgPattern(r io.Reader, mp MessagePattern) error {
+
+	// Read and check version.
+	version := [1]byte{}
+	if _, err := r.Read(version[:]); err != nil {
+		return err
+	}
+
+	if version[0] != h.version {
+		return fmt.Errorf("received unexpected handshake version: %v",
+			version[0])
+	}
+
+	// Read Tokens.
+	if err := h.readTokens(r, mp.Tokens); err != nil {
+		return err
 	}
+
+	// Read the payload data.
+	switch h.version {
+	case HandshakeVersion:
+		payloadSize := 0
+		if mp.ActNum == act2 {
+			payloadSize = ActTwoPayloadSize
+		}
+
+		cipherText := make([]byte, payloadSize+macSize)
+		_, err := r.Read(cipherText[:])
+		if err != nil {
+			return err
+		}
+
+		payload, err := h.DecryptAndHash(cipherText[:])
+		if err != nil {
+			return err
+		}
+
+		// If the payload is a non-zero length, then we'll assume it's
+		// the auth data and attempt to fully decode it.
+		if len(payload) == 0 {
+			return nil
+		}
+
+		payloadLen := binary.BigEndian.Uint16(
+			payload[:lengthHeaderSize],
+		)
+		authData := make([]byte, payloadLen)
+
+		payloadReader := bytes.NewReader(payload[lengthHeaderSize:])
+		if _, err := payloadReader.Read(authData); err != nil {
+			return err
+		}
+
+		h.receivedPayload = authData
+	default:
+		return fmt.Errorf("unknown handshake version: %v", h.version)
+	}
+
+	return nil
+}
+
+// writeToken processes a series of Tokens and either writes a public key
+// (unencrypted, encrypted or masked) to the writer or performs a ECDH
+// operation.
+func (h *handshakeState) writeTokens(tokens []Token, w io.Writer) error {
```

Comment:
Nice, had always wanted to refactor things to more generic, but didn't have another handshake we were using at the time, this sets us up nicely to add the secondary longer term handshake.